### ISSUE TITLE
aref for selecting arbitrary elements

### ISSUE THREAD

**Issue by digikar99** [numcl/numcl#32]
Basically, this

```python
a = np.array([[1,2,3,4],[5,6,7,8]])
a[:, [0, 2,3]] # returns array([[1,3,4], [5,7,8]])
```

Is there a numcl equivalent? Because `(aref a t '(0 2 3))` has a different semantics translating to `a[:, 0:2:3]`. I find needing the above slicing. There is the [select:select](https://github.com/Symbolics/select), but this returns a simple-array.

May be a configuration variable may be included to let users choose between the two notations.

PS: I raised quite a few issues today. Lastly, thank you for all the work!

**Comment by guicho271828** [comment]
Interesting. Do you assume `'(0 2 3)` to be mainly a constant (like in this example) or mainly be a variable? For example,
```
(defun select (array selected)
  (aref array t selected))
```
where `selected` could be `(0 2 3)`, or `(0 1 3)`?
The latter would be written better with a combination of `concatenate` and `mapcar`, but I understand the need for the former. 

What do you think about candidates like this?
```
(aref a t '(:select 0 2 3))
```

**Comment by guicho271828** [comment]
For the implementation convenience, `(aref a t '(:select (0 2 3)))` would be better --- I can possibly implement it in einsum too.

**Comment by digikar99** [comment]
>Do you assume '(0 2 3) to be mainly a constant (like in this example) or mainly be a variable?

I think, I'd rather have it as a variable.

If both are possible simultaneously, I find `(numcl:aref a t #(0 2))` rather wanting, again in line with numpy semantics:

```py
idx = np.array([0,2]) # could be just the python list [0, 2]
a[:, idx]
```

However, I don't understand [this](https://docs.scipy.org/doc/numpy/reference/arrays.indexing.html) - the topics on advanced slicing is quite a lot to take in for me; and I do wonder how many people even use all of that. It does suggest that the array used for indexing (`#(0 2)` above) could also be multidimensional - I'd guess this'd adds quite some complexity. And I'm not sure if anything such as advanced indexing (beyond single-dimensional arrays) is something useful or something obscure. 

But, there also is `take` and `where`! 

The least "features" I can find useful is using integer or boolean vectors for indexing, I haven't yet encountered a use case in my very limited experience with ML beyond that.

**Comment by guicho271828** [comment]
> using integer or boolean vectors for indexing

Indeed, I also use this feature out of necessity in one of my machine learning projects, so I understand. Maybe arrays as the index objects might make more sense.
